waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim

```diff
@@ -952,67 +962,85 @@ proc appendRLNProof*(rlnPeer: WakuRLNRelay, msg: var WakuMessage,
+    let isSuccessful = rlnPeer.insertMembers(startingIndex, members.mapIt(it.idComm))
     if isSuccessful.isErr():
-      return err("failed to add a new member to the Merkle tree")
+      return err("failed to add new members to the Merkle tree")
     else:
-      debug "new member added to the Merkle tree", pubkey=pubkey, index=index
+      debug "new members added to the Merkle tree", pubkeys=members.mapIt(it.idComm.inHex) , startingIndex=startingIndex
       debug "acceptable window", validRoots=rlnPeer.validMerkleRoots.mapIt(it.inHex)
-      let membershipIndex = index.toMembershipIndex()
-      if rlnPeer.lastSeenMembershipIndex != membershipIndex + 1:
-        warn "membership index gap, may have lost connection", gap = membershipIndex - rlnPeer.lastSeenMembershipIndex
-      rlnPeer.lastSeenMembershipIndex = membershipIndex
+      let lastIndex = members[0].index + members.len.uint - 1
+      let indexGap = lastIndex - rlnPeer.lastSeenmembershipIndex
+      if indexGap > 1.uint:
+        warn "membership index gap, may have lost connection", lastIndex, currIndex=rlnPeer.lastSeenMembershipIndex, indexGap = indexGap
+      rlnPeer.lastSeenMembershipIndex = lastIndex
+      rlnPeer.lastProcessedBlock = blockNumber
+      debug "last processed block", blockNumber = blockNumber
       return ok()
   return handler
 
-proc subscribeToMemberRegistrations(web3: Web3, 
-                                    contractAddress: Address,
-                                    fromBlock: string = "0x0",
-                                    handler: GroupUpdateHandler): Future[Subscription] {.async, gcsafe.} =
-  ## subscribes to member registrations, on a given membership group contract
-  ## `fromBlock` indicates the block number from which the subscription starts
-  ## `handler` is a callback that is called when a new member is registered
-  ## the callback is called with the pubkey and the index of the new member
-  ## TODO: need a similar proc for member deletions
-  var contractObj = web3.contractSender(MembershipContract, contractAddress)
-
-  let onMemberRegistered = proc (pubkey: Uint256, index: Uint256) {.gcsafe.} =
-    debug "onRegister", pubkey = pubkey, index = index
-    var groupUpdateRes: RlnRelayResult[void]
-    try:
-      groupUpdateRes = handler(pubkey, index)
-    except Exception as err:
-      error "failed to handle group update", err = err.msg
-    if groupUpdateRes.isErr():
-      error "Error handling new member registration", err=groupUpdateRes.error()
-
-  let onError = proc (err: CatchableError) =
-    error "Error in subscription", err=err.msg
-
-  return await contractObj.subscribe(MemberRegistered,
-                                     %*{"fromBlock": fromBlock, "address": contractAddress},
-                                     onMemberRegistered,
-                                     onError)
+proc parse*(event: type MemberRegistered, 
```

Comment:
Please also add a comment to each line of this code explaining what is happeing.

Reply:
Addressed in [8b0ff46](https://github.com/status-im/nwaku/pull/1349/commits/8b0ff463b4b8f307d33475e365ca1d0c389e39d7)